### PR TITLE
29 extend polynomial dfs

### PR DESCRIPTION
**Pull request by ETatuzova**

### CODE REVIEW

File: include/nil/crypto3/math/polynomial/polynomial_dfs.hpp [hunk 1]
```diff
@@ -36,6 +36,8 @@
 namespace nil {
     namespace crypto3 {
         namespace math {
+            //size_t __global_from_coefficients_counter_test = 0;
```
Comment (nkaskov) [inline include/nil/crypto3/math/polynomial/polynomial_dfs.hpp:39]: Shall we remove these lines?

File: include/nil/crypto3/math/polynomial/polynomial_dfs.hpp [hunk 2]
```diff
@@ -540,48 +611,73 @@ namespace nil {
                      typename = typename std::enable_if<detail::is_field_element<FieldValueType>::value>::type>
             polynomial_dfs<FieldValueType, Allocator> operator+(const polynomial_dfs<FieldValueType, Allocator>& A,
                                                             const FieldValueType& B) {
-
-                return A + polynomial_dfs<FieldValueType>(0, A.size(), B);
+//                return A + polynomial_dfs<FieldValueType>(0, A.size(), B);
```
Comment (nkaskov) [inline include/nil/crypto3/math/polynomial/polynomial_dfs.hpp:614]: Here and below: commented lines. Do we want to keep it?

**Review by nkaskov — CHANGES_REQUESTED** [review]

File: include/nil/crypto3/math/polynomial/polynomial_dfs.hpp [hunk 3]
```diff
@@ -411,6 +413,31 @@ namespace nil {
                     return result;
                 }
 
+                /**
+                 * Computes the standard polynomial addition, polynomial A += polynomial B, and stores result in
+                 * polynomial C.
```
Comment (NickVolynkin) [inline include/nil/crypto3/math/polynomial/polynomial_dfs.hpp:418]: It's slightly more readable if you wrap like this
```suggestion
                 * Computes the standard polynomial addition, polynomial A += polynomial B,
                 * and stores result in polynomial C.
```

File: include/nil/crypto3/math/polynomial/polynomial_dfs.hpp [hunk 4]
```diff
@@ -438,6 +465,30 @@ namespace nil {
                     return result;
                 }
 
+                /**
+                 * Computes the standard polynomial addition, polynomial A -= polynomial B, and stores result in
+                 * polynomial C.
```
Comment (NickVolynkin) [inline include/nil/crypto3/math/polynomial/polynomial_dfs.hpp:470]: `-=` is substraction, not addition. And the same note about wrapping as above.

File: include/nil/crypto3/math/polynomial/polynomial_dfs.hpp [hunk 5]
```diff
@@ -438,6 +465,30 @@ namespace nil {
...
+                    std::transform(this->begin(), this->end(), other.begin(), this->begin(), std::minus<FieldValueType>());
+                    return *this;
+                }
+
+                polynomial_dfs operator-=(const FieldValueType& c) {
+                    for( auto it = this->begin(); it!=this->end(); it++) *it -= c;
+                    return *this;
+                }
                 /**
                  * Perform the multiplication of two polynomials, polynomial A * polynomial B, and stores result in
```
Comment (NickVolynkin) [inline include/nil/crypto3/math/polynomial/polynomial_dfs.hpp:493]: Words don't agree grammatically:

> Perform... and stores

I like the infinitive option more:

```suggestion
                 * Perform the multiplication of two polynomials: polynomial A * polynomial B,
                 * and store the result in polynomial C.
```

File: include/nil/crypto3/math/polynomial/polynomial_dfs.hpp [hunk 6]
```diff
@@ -540,48 +611,73 @@ namespace nil {
...
+                return result;
             }
 
             template<typename FieldValueType, typename Allocator = std::allocator<FieldValueType>,
                      typename = typename std::enable_if<detail::is_field_element<FieldValueType>::value>::type>
             polynomial_dfs<FieldValueType, Allocator> operator*(const FieldValueType& A,
                                                             const polynomial_dfs<FieldValueType, Allocator>& B) {
-
-                return polynomial_dfs<FieldValueType>(0, B.size(), A) * B;
+//                return polynomial_dfs<FieldValueType>(0, B.size(), A) * B;
```
Comment (NickVolynkin) [inline include/nil/crypto3/math/polynomial/polynomial_dfs.hpp:675]: same as above

File: include/nil/crypto3/math/polynomial/polynomial_dfs.hpp [hunk 7]
```diff
@@ -540,48 +611,73 @@ namespace nil {
...
+                return result;
             }
 
             template<typename FieldValueType, typename Allocator = std::allocator<FieldValueType>,
                      typename = typename std::enable_if<detail::is_field_element<FieldValueType>::value>::type>
             polynomial_dfs<FieldValueType, Allocator> operator-(const polynomial_dfs<FieldValueType, Allocator>& A,
                                                             const FieldValueType& B) {
-
-                return A - polynomial_dfs<FieldValueType>(0, A.size(), B);
+//                return A - polynomial_dfs<FieldValueType>(0, A.size(), B);
```
Comment (NickVolynkin) [inline include/nil/crypto3/math/polynomial/polynomial_dfs.hpp:638]: this line probably should be removed

**Review by nkaskov — APPROVED** [review]

**Pull request merged**